sui/gas-service.js

```diff
@@ -0,0 +1,220 @@
+        ],
+    });
+
+    const receipt = await broadcast(client, keypair, tx);
+
+    printInfo('Gas paid', receipt.digest);
+}
+
+async function addGas(keypair, client, gasServiceConfig, args, options) {
+    const walletAddress = keypair.toSuiAddress();
+
+    const gasServicePackageId = gasServiceConfig.address;
+
+    const refundAddress = options.refundAddress || walletAddress;
+    const params = options.params || '0x';
+
+    const [messageId, amount] = args;
+
+    const atomicAmount = getAtomicAmount(amount);
+
+    const tx = new TransactionBlock();
+    const [coin] = tx.splitCoins(tx.gas, [atomicAmount]);
+
+    tx.moveCall({
+        target: `${gasServicePackageId}::gas_service::add_gas`,
+        arguments: [
+            tx.object(gasServiceConfig.objects.GasService),
+            coin, // Coin<SUI>
+            tx.pure(bcs.string().serialize(messageId).toBytes()), // Message ID for the contract call
+            tx.pure.address(refundAddress), // Refund address
+            tx.pure(bcs.vector(bcs.u8()).serialize(arrayify(params)).toBytes()), // Params
+        ],
+    });
+
+    const receipt = await broadcast(client, keypair, tx);
+
+    printInfo('Gas added', receipt.digest);
+}
+
+async function collectGas(keypair, client, gasServiceConfig, args, options) {
+    const walletAddress = keypair.toSuiAddress();
+
+    const gasServicePackageId = gasServiceConfig.address;
+    const gasServiceObjectId = gasServiceConfig.objects.GasService;
+
+    const [amount] = args;
+    const receiver = options.receiver || walletAddress;
+
+    const atomicAmount = getAtomicAmount(amount);
+
+    const bytes = await getBcsBytesByObjectId(client, gasServiceObjectId);
+    const { balance: gasServiceBalance } = gasServiceStruct.parse(bytes);
+
+    // Check if the gas service balance is sufficient
+    if (gasServiceBalance < atomicAmount) {
+        printError('Insufficient gas service balance', `${getFormattedAmount(gasServiceBalance)} < ${getFormattedAmount(atomicAmount)}`);
+        return;
+    }
+
+    const tx = new TransactionBlock();
+
+    tx.moveCall({
+        target: `${gasServicePackageId}::gas_service::collect_gas`,
+        arguments: [
+            tx.object(gasServiceConfig.objects.GasService),
+            tx.object(gasServiceConfig.objects.GasCollectorCap),
+            tx.pure.address(receiver), // Receiver address
+            tx.pure.u64(atomicAmount), // Amount
+        ],
+    });
+
+    const receipt = await broadcast(client, keypair, tx);
+
+    printInfo('Gas collected', receipt.digest);
+}
+
+async function refund(keypair, client, gasServiceConfig, args, options) {
+    const walletAddress = keypair.toSuiAddress();
+
+    const gasServicePackageId = gasServiceConfig.address;
+    const gasServiceObjectId = gasServiceConfig.objects.GasService;
+
+    const [messageId, amount] = args;
+    const receiver = options.receiver || walletAddress;
+
+    const atomicAmount = getAtomicAmount(amount);
```

Comment:
commander provides option processing, so we can do the conversion there instead of duplicating it
https://github.com/tj/commander.js?tab=readme-ov-file#custom-option-processing

we can use this for validation as well, and any other processing

Reply:
amount is positional param, so custom option processing doesn't work. If we convert amount into option, then it's currently hard to define default value for e.g. addGas function.

Reply:
hmm, no need to define a default. it can just be a required option. i prefer reusing the validation over keeping it as a positional arg